lib/queue.js

```diff
@@ -588,45 +586,43 @@ class Queue extends Emitter {
       const delay = strategy ? strategy(job) : -1;
       if (delay < 0) {
         job.status = 'failed';
-        if (this.settings.removeOnFailure) {
-          multi.hdel(this.toKey('jobs'), job.id);
-        } else {
-          multi.hset(this.toKey('jobs'), job.id, job.toData());
-          multi.sadd(this.toKey('failed'), job.id);
-        }
+        removeJob = this.settings.removeOnFailure;
       } else {
         job.options.retries -= 1;
         job.status = 'retrying';
         jobEvent.event = 'retrying';
-        multi.hset(this.toKey('jobs'), job.id, job.toData());
-        if (delay === 0) {
-          multi.lpush(this.toKey('waiting'), job.id);
-        } else {
-          const time = Date.now() + delay;
-          multi
-            .zadd(this.toKey('delayed'), time, job.id)
-            .publish(this.toKey('earlierDelayed'), time);
+        if (delay !== 0) {
+          delayedTime = Date.now() + delay;
         }
       }
     } else {
       job.status = 'succeeded';
-      if (this.settings.removeOnSuccess) {
-        multi.hdel(this.toKey('jobs'), job.id);
-      } else {
-        multi.hset(this.toKey('jobs'), job.id, job.toData());
-        multi.sadd(this.toKey('succeeded'), job.id);
-      }
-    }
-
-    if (this.settings.sendEvents) {
-      multi.publish(this.toKey('events'), JSON.stringify(jobEvent));
+      removeJob = this.settings.removeOnSuccess;
     }
 
     const result = err || data;
 
-    return helpers
-      .callAsync((done) => multi.exec(done))
-      .then(() => [status, result]);
+    const promise = this._evalScript(
+      'finishJob',
+      9,
+      this.toKey('active'),
+      this.toKey('waiting'),
+      this.toKey('stalling'),
+      this.toKey('failed'),
+      this.toKey('delayed'),
+      this.toKey('earlierDelayed'),
+      this.toKey('succeeded'),
+      this.toKey('events'),
+      this.toKey('jobs'),
+      job.id,
+      removeJob ? 'true' : '',
+      job.toData(),
+      job.status,
+      delayedTime || '',
+      this.settings.sendEvents && JSON.stringify(jobEvent)
```

Comment:
does this work if we don't pass in an empty string?
```suggestion
      this.settings.sendEvents ? JSON.stringify(jobEvent) : ''
```